# SKU Creation

# Business Problem:
As per my conversation with Allen Shimon, StrobesnMore needs a merge of their products and product
options so that every possible combination of options has a unique skew. The challenge that their
website currently faces is that their inventory lookup requires visiting multiple links and clicking into
mulƟple dropdowns, making their current list ineffective at looking up product quickly.

`Example:
Product 123 has 2 options with multiple types.`
- Option 1(product feature): Magnetic VS Permanent
- Option 2(color): Blue VS Green

CURRENTLY: Product 123 only has 1 skew (123)
After the complete skew list is created, the following skews will be provided:
- 123MBLUE
- 123PBLUE
- 123MGREEN
- 123PGREEN


Data Understanding
Two csv files were provided to demonstrate the need for this service. There is a csv with a full list of products and another with a list of product options. The goal will be to provide a query merging these two csv files so that each product has a skew associated with every combination of options associated with it. This process will require “digging into” the data and coding a query to pull the list of products and merge all their combination of options. To ensure quality, an understanding of the data options and their products will be required. In the product options list, a few observations have been made and need to be confirmed…

1. It is assumed that each product can only have one unique color.
2. The optionid is the identifier for the option categories for each product. It is assumed that each product can only have one unique option (for example, cannot be both magnetic and permanent) per optionid.
3. The catalogid, NOT the productid in the csv appears to the proper unique identifier for each product.
- `**Edit** SKUs will be created with the productid, not the catalogid, along with the partnumbers.`
4. Client should list if there is a preffered format for SKU creation(color first, capitals, etc)
- `**Edit** Client has issued that he needs the SKUs to be an exact match with the website.`

# Google Colab

This code wad executed in google colab, so the block of code below is the libraries needed to mount the notebook to the users google drive. Depending on the path to your file, the users may be different. 

In [1]:
# from google.colab import drive
# drive.mount('/content/drive')


# Libraries Used 

Below are the libraries needed to run the script. 

- The pandas library was used to manipulate the initial spreadsheet. 
- The itertools product was used to compile the permutations (AKA the SKUs) for all the products. 

In [2]:
import pandas as pd
import numpy as np
from itertools import product


pd.set_option('display.max_rows', None)
pd.set_option("display.max_colwidth", 100)

### Functions

Here we have the functions that will obtain the dataframe of SKUs. The process is as follows:

1) get_tuples_for_all_categories, get_tuples_for_all_labels:

These functions get us the partnumbers and featurenames for all the categories associated with each product. The function returns tuples of the partnumber and the sorting number, which in theory is how the partnumbers should be ordered for the SKUs.

2) generate_combinations:

The combination function generates all combinations of the partnumbers, with consideration of order using the 'sorting' column from the dataframe. The functions creates lists of all the possible outcomes and then returns the largest list as some sets of SKUs come out with 1, 2, or 3 options(more options can be added to the function if need be).

3) create_combo_dataframe:

This is the function that puts all these parts together. It iterates through the dataframe and gets all the labels, partnumbers, and combinations and then appends them to a new dataframe composed of the catalogid, productid, the SKU, and the description of the SKU.

4) get_max_unique_partnumbers
This function, get_max_unique_partnumbers, aims to identify the catalog ID (CatalogID) within a given DataFrame (df) that has the maximum number of unique sorting values. It begins by specifying the column names for the catalog ID and the sorting values. Then, it iterates through each unique catalog ID in the DataFrame, filters the DataFrame to obtain data for each catalog ID, and calculates the number of unique sorting values for each catalog ID. During this process, it tracks the catalog ID with the maximum unique sorting values and the corresponding count. Finally, it prints out or returns the catalog ID with the maximum number of unique sorting values along with the count. This function provides a way to identify which catalog ID has the most diverse range of sorting values, which may be useful for various analytical or data exploration purposes.

In [3]:

def get_tuples_for_all_categories(df):
    # Extract unique category values from the 'cat_w_option' column
    unique_categories = df['cat_w_option'].unique()
    # Initialize an empty list to store tuples of data for each category
    data_by_category = []

    # Iterate through each unique category value
    for category_value in unique_categories:
        # Extract pairs of 'OptionPartNumber' and 'OptionSorting' for the current category
        partnumber_sorting_pairs = list(zip(
            df[df['cat_w_option'] == category_value]['OptionPartNumber'],
            df[df['cat_w_option'] == category_value]['OptionSorting']
        ))
        # Extend the data_by_category list with the extracted pairs
        data_by_category.extend(partnumber_sorting_pairs)

    # Return the list containing tuples of data for all categories
    return data_by_category


def get_tuples_for_all_labels(df):
    # Extract unique category values from the 'cat_w_option' column
    unique_categories = df['cat_w_option'].unique()
    # Initialize an empty list to store tuples of data for each category
    data_by_category = []

    # Iterate through each unique category value
    for category_value in unique_categories:
        # Extract pairs of 'OptionName' and 'OptionSorting' for the current category
        partnumber_sorting_pairs = list(zip(
            df[df['cat_w_option'] == category_value]['OptionName'],
            df[df['cat_w_option'] == category_value]['OptionSorting']
        ))
        # Extend the data_by_category list with the extracted pairs
        data_by_category.extend(partnumber_sorting_pairs)

    # Return the list containing tuples of data for all categories
    return data_by_category



def generate_combinations(input_list):
    # Initialize empty lists to store elements for each unique value
    l1 = []
    l2 = []
    l3 = []
    l4 = []
    l5 = []
    l6 = []
    l7 = []
    l8 = []
    l9 = []
    l10 = []
    l11 = []
    l12 = []

    # Extract values (second elements) from input_list and sort them
    values = [i[1] for i in input_list]
    values = sorted(values)
    # Get unique values from the sorted list
    unique_values = list(set(values))

    # Iterate through input_list
    for i in input_list:
        # Distribute elements from input_list into appropriate lists based on their second element
        if i[1] == unique_values[0]:
            l1.append(i[0])
        elif i[1] == unique_values[1]:
            l2.append(i[0])
        elif i[1] == unique_values[2]:
            l3.append(i[0])
        elif i[1] == unique_values[3]:
            l4.append(i[0])
        elif i[1] == unique_values[4]:
            l5.append(i[0])
        elif i[1] == unique_values[5]:
            l6.append(i[0])
        elif i[1] == unique_values[6]:
            l7.append(i[0])
        elif i[1] == unique_values[7]:
            l8.append(i[0])
        elif i[1] == unique_values[8]:
            l9.append(i[0])
        elif i[1] == unique_values[9]:
            l10.append(i[0])
        elif i[1] == unique_values[10]:
            l11.append(i[0])
        elif i[1] == unique_values[11]:
            l12.append(i[0])

    # Generate combinations based on the number of unique values
    if len(unique_values) == 1:
        perm_1 = l1
        return perm_1
    elif len(unique_values) == 2:
        perm_1_2 = list(product(l1, l2))
        return perm_1_2
    elif len(unique_values) == 3:
        perm_1_2_3 = list(product(l1, l2, l3))
        return perm_1_2_3
    elif len(unique_values) == 4:
        perm_1_2_3_4 = list(product(l1, l2, l3, l4))
        return perm_1_2_3_4
    elif len(unique_values) == 5:
        perm_1_2_3_4_5 = list(product(l1, l2, l3, l4, l5))
    elif len(unique_values) == 6:
        perm_1_2_3_4_5_6 = list(product(l1, l2, l3, l4, l5, l6))
        return perm_1_2_3_4_5_6
    elif len(unique_values) == 7:
        perm_1_2_3_4_5_6_7 = list(product(l1, l2, l3, l4, l5, l6, l7))
        return perm_1_2_3_4_5_6_7
    elif len(unique_values) == 8:
        perm_1_2_3_4_5_6_7_8 = list(product(l1, l2, l3, l4, l5, l6, l7, l8))
        return perm_1_2_3_4_5_6_7_8
    elif len(unique_values) == 9:
        perm_1_2_3_4_5_6_7_8_9 = list(product(l1, l2, l3, l4, l5, l6, l7, l8, l9))
        return perm_1_2_3_4_5_6_7_8_9
    elif len(unique_values) == 10:
        perm_1_2_3_4_5_6_7_8_9_10 = list(product(l1, l2, l3, l4, l5, l6, l7, l8, l9, l10))
        return perm_1_2_3_4_5_6_7_8_9_10
    elif len(unique_values) == 11:
        perm_1_2_3_4_5_6_7_8_9_10_11 = list(product(l1, l2, l3, l4, l5, l6, l7, l8, l9, l10, l11))
        return perm_1_2_3_4_5_6_7_8_9_10_11
    elif len(unique_values) == 12:
        perm_1_2_3_4_5_6_7_8_9_10_11_12 = list(product(l1, l2, l3, l4, l5, l6, l7, l8, l9, l10, l11, l12))
        return perm_1_2_3_4_5_6_7_8_9_10_11_12



def create_combo_dataframes(df):
    # DataFrame df with columns 'catalogid', 'productid', 'partnumber', 'sorting', 'labels'
    unique_values = df['CatalogID'].unique()

    # Initialize an empty DataFrame to store the results
    combined_df = pd.DataFrame()

    # Iterate through unique values in the specified column
    for value in unique_values:
        # Create a subset DataFrame based on the column value
        subset_df = df[df['CatalogID'] == value]

        # Apply your functions to the subset DataFrame
        result_PN = get_tuples_for_all_categories(subset_df)
        result_labels = get_tuples_for_all_labels(subset_df)
        combination_PN = generate_combinations(result_PN)
        combination_labels = generate_combinations(result_labels)

        # Concatenate the results for each value
        combo_PN = pd.DataFrame(combination_PN).sum(axis=1)
        combo_labels = pd.DataFrame(combination_labels).apply(lambda row: '_'.join(row), axis=1)

        # Add 'productid' column
        name_column = pd.Series(subset_df['Name'].unique(), name='Name')
        product_id_column = pd.Series(subset_df['SKU'].unique(), name='ProductID')
        catalog_id_column = pd.Series(subset_df['CatalogID'].unique(), name='CatalogID')

        # Fill NaN values in combo_PN
        combo_PN = combo_PN.fillna('')

        # Convert float values to strings
        combo_PN = combo_PN.astype(str)
        product_id_column = product_id_column.astype(str)

        # Concatenate all columns
        combo_df = pd.concat([catalog_id_column, product_id_column, combo_PN,name_column, combo_labels], axis=1)

        # Concatenate to the overall DataFrame
        combined_df = pd.concat([combined_df, combo_df])

    # Set columns outside the loop
    combined_df.columns = ['CatalogID', 'ProductID', 'SKUs','Name', 'Labels']
    combined_df = combined_df.fillna(method='ffill')
    combined_df['SKU'] = combined_df['ProductID'] + combined_df['SKUs']
    final_df = combined_df[['CatalogID','ProductID', 'SKU','Name', 'Labels']]

    return final_df

def get_max_unique_partnumbers(df):
    # Replace 'catalogid_column_name' with the actual column name containing catalogid
    # Replace 'sorting_column_name' with the actual column name containing sorting values

    catalogid_column_name = 'CatalogID'
    sorting_column_name = 'OptionSorting'

    # Create variables to store the catalogid with max unique values and the max count
    max_catalogid = None
    max_unique_count = 0

    # Iterate through unique catalogids in the DataFrame
    for catalogid in df[catalogid_column_name].unique():
        # Filter DataFrame for the current catalogid
        catalogid_data = df[df[catalogid_column_name] == catalogid]

        # Count the number of unique sorting values
        unique_sorting_count = len(catalogid_data[sorting_column_name].unique())

        # Check if the current catalogid has more unique values than the current max
        if unique_sorting_count > max_unique_count:
            max_catalogid = catalogid
            max_unique_count = unique_sorting_count

    # Print or use the max_catalogid and max_unique_count as needed
    print(f"The catalogid with the maximum number of unique sorting values is {max_catalogid} with {max_unique_count} unique values.")





# Reading DataFrame


This block of code performs several data preprocessing steps on an Excel dataset containing options information. 

1. Reads the Excel file located at the specified path into a Pandas DataFrame. 
2. Converts certain columns like 'OptionSetID', 'CatalogID', 'OptionPartNumber', and 'OptionName' to string type to ensure consistent data type handling. The values in the 'OptionSorting' column are incremented by 1. 
3. A new column 'cat_w_option' is created by concatenating 'CatalogID' and 'OptionSetID'. Any 'nan' values in the 'OptionPartNumber' column are replaced with an empty string. 
4. A subset DataFrame named 'df_test' is created for testing purposes, containing rows from index 17 to 60. 
5. ain DataFrame is filtered to exclude rows where the 'Hidden' column has a value of 1, and the first few rows of the resulting DataFrame are displayed. Overall, this block of code prepares the dataset for further analysis by ensuring data consistency, creating new columns, and filtering out irrelevant rows.

In [4]:
# Define the file path
path = "All_Options.xlsx"

# Read the Excel file into a DataFrame
df_options_ = pd.read_excel(path)

# Convert certain columns to string type
df_options_['OptionSetID'] = df_options_['OptionSetID'].astype(str)
df_options_['CatalogID'] = df_options_['CatalogID'].astype(str)
df_options_['OptionPartNumber'] = df_options_['OptionPartNumber'].astype(str)
df_options_['OptionName'] = df_options_['OptionName'].astype(str)

# Increment values in the 'OptionSorting' column by 1
df_options_['OptionSorting'] = df_options_['OptionSorting'] + 1

# Create a new column 'cat_w_option' by concatenating 'CatalogID' and 'OptionSetID'
df_options_['cat_w_option'] = df_options_['CatalogID'] + '_' + df_options_['OptionSetID']

# Replace 'nan' values in 'OptionPartNumber' with empty string
df_options_['OptionPartNumber'] = df_options_['OptionPartNumber'].apply(lambda x: '' if x=='nan' else x)

# Create a subset DataFrame for testing purposes
df_test = df_options_.iloc[17:61]

# Filter the main DataFrame to include only rows where 'Hidden' column is 0
df_options_ = df_options_[df_options_['Hidden'] == 0]

# Display the first few rows of the DataFrame
df_options_.head()


Unnamed: 0 CatalogID      SKU                                  Name  \
0           4      3474  11.1002  Sho Me Universal On/Off/Flash Switch   
1           5      3474  11.1002  Sho Me Universal On/Off/Flash Switch   
2           6      3475   11.120  Sho Me Low Profile LED Mini Lightbar   
3           7      3475   11.120  Sho Me Low Profile LED Mini Lightbar   
4           8      3475   11.120  Sho Me Low Profile LED Mini Lightbar   

   NotForSale   Hide                 OptionName OptionPartNumber OptionSetID  \
0       False  False         Positive Momentary              .PM        4114   
1       False  False  Negative/Ground Momentary             .GSM        4114   
2       False  False                   Magnetic            0.008        2457   
3       False  False                  Permanent                0        2457   
4       False  False                       Blue              -BB        2458   

   OptionSorting  Hidden cat_w_option  
0              2       0    3474_4114  
1              2       0    3474_4114  
2              1       0    3475_2457  
3              1       0    3475_2457  
4              2       0    3475_2458

In [5]:
# path = "/content/drive/My Drive/Colab Notebooks/Grades/allproducts.xlsx"
# df_products = pd.read_excel(path)



# df_products['OptionSetID'] = df_products['OptionSetID'].astype(str)
# df_products['CatalogID'] = df_products['CatalogID'].astype(str)
# df_products['OptionPartNumber'] = df_products['OptionPartNumber'].astype(str)
# df_products['OptionName'] = df_products['OptionName'].astype(str)
# df_products['OptionSorting'] = df_products['OptionSorting'] + 1
# df_products['cat_w_option'] =  df_products['CatalogID'] + '_' + df_products['OptionSetID']
# df_products['OptionPartNumber'] = df_products['OptionPartNumber'].apply(lambda x: '' if x=='nan' else x)
# # df_test = df_products.iloc[17:61]
# df_products = df_products[df_products['Hidden'] == 0]


# df_products

# Checking for NotForSale and NotOnWebsite Products.

During the scraping of the product data, it was discovered that some of the products from the API are listed as not for sale or not on the website. We chose not to include these products in the final SKU list, but kept the list for inventory purposes. 

## Filtering by NotForSale

In [6]:
NotForSale = df_options_[df_options_['NotForSale'] == True]

## Checking 'Hide' Values

If a product is listed as 'Hide' = True, it is off the website. If listed as False, then it is on the website. 

In [7]:
df_options_['Hide'].unique()

array([False])

## Checking 'NotForSale' Values

If a product is listed as 'NotForSale' = True, it is off the website. If listed as False, then it is on the website. 

In [8]:
df_options_['NotForSale'].unique()

array([False,  True])

## All catalogIDs of products not for sale, after filtering. 

In [9]:
NotForSale['CatalogID'].unique()

array(['5862', '5888', '6743'], dtype=object)

In [10]:
NotForSale.head(10)

Unnamed: 0 CatalogID    SKU                Name  NotForSale   Hide  \
1816         466      5862  PB400  Setina Push Bumper        True  False   
1817         467      5862  PB400  Setina Push Bumper        True  False   
1818         468      5862  PB400  Setina Push Bumper        True  False   
1819         469      5862  PB400  Setina Push Bumper        True  False   
1820         470      5862  PB400  Setina Push Bumper        True  False   
1822         472      5862  PB400  Setina Push Bumper        True  False   
1823         473      5862  PB400  Setina Push Bumper        True  False   
1824         474      5862  PB400  Setina Push Bumper        True  False   
1825         475      5862  PB400  Setina Push Bumper        True  False   
1826         476      5862  PB400  Setina Push Bumper        True  False   

                          OptionName OptionPartNumber OptionSetID  \
1816    Ford Interceptor Sedan 2012+            ITS12        4601   
1817  Ford Interceptor Utility 2016+            ITU16        4601   
1818           Ford Expedition 2015+            EPD15        4601   
1819                Ford F-150 2015+        FDT15F150        4601   
1820            Ford F-250-550 2018+        FDT17F250        4601   
1822     Chevrolet Caprice 2011-2017            CAP11        4601   
1823          Chevrolet Impala 2014+            IMP06        4601   
1824           Chevrolet Tahoe 2015+            TAH15        4601   
1825  Chevrolet Silverado 1500 2014+        CHT141500        4601   
1826             Dodge Charger 2015+            CGR15        4601   

      OptionSorting  Hidden cat_w_option  
1816              9       0    5862_4601  
1817              9       0    5862_4601  
1818              9       0    5862_4601  
1819              9       0    5862_4601  
1820              9       0    5862_4601  
1822              9       0    5862_4601  
1823              9       0    5862_4601  
1824              9       0    5862_4601  
1825              9       0    5862_4601  
1826              9       0    5862_4601

# Removing Items Not for Sale

In [11]:
df_options_ = df_options_[df_options_['NotForSale'] == False]

# Grabbing Columns

### Checking length of dataframe after filtering by products we intend to get SKUs for

In [12]:
len(df_options_)

3391

In [13]:
df_options_.columns

Index(['Unnamed: 0', 'CatalogID', 'SKU', 'Name', 'NotForSale', 'Hide',
       'OptionName', 'OptionPartNumber', 'OptionSetID', 'OptionSorting',
       'Hidden', 'cat_w_option'],
      dtype='object')

## Removing the 'Unnamed: 0' Column

In [14]:
df_options_ = df_options_[['CatalogID','SKU','OptionName','Name','OptionPartNumber','OptionSetID','OptionSorting','cat_w_option']]

In [15]:
df_options_.head()

CatalogID      SKU                 OptionName  \
0      3474  11.1002         Positive Momentary   
1      3474  11.1002  Negative/Ground Momentary   
2      3475   11.120                   Magnetic   
3      3475   11.120                  Permanent   
4      3475   11.120                       Blue   

                                   Name OptionPartNumber OptionSetID  \
0  Sho Me Universal On/Off/Flash Switch              .PM        4114   
1  Sho Me Universal On/Off/Flash Switch             .GSM        4114   
2  Sho Me Low Profile LED Mini Lightbar            0.008        2457   
3  Sho Me Low Profile LED Mini Lightbar                0        2457   
4  Sho Me Low Profile LED Mini Lightbar              -BB        2458   

   OptionSorting cat_w_option  
0              2    3474_4114  
1              2    3474_4114  
2              1    3475_2457  
3              1    3475_2457  
4              2    3475_2458

# Checking Datatypes

Here we are making sure the dataframe has the correct datatypes for the functions to run.

In [16]:
df_options_.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3391 entries, 0 to 3766
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   CatalogID         3391 non-null   object
 1   SKU               3391 non-null   object
 2   OptionName        3391 non-null   object
 3   Name              3391 non-null   object
 4   OptionPartNumber  3391 non-null   object
 5   OptionSetID       3391 non-null   object
 6   OptionSorting     3391 non-null   int64 
 7   cat_w_option      3391 non-null   object
dtypes: int64(1), object(7)
memory usage: 238.4+ KB


# Checking that the max product number of options is up to 8. 

Client has agreed that the cutoff for number of options to be founs for SKUs to be 8. This function checks for the max number of options for the products.

In [17]:
get_max_unique_partnumbers(df_options_)

The catalogid with the maximum number of unique sorting values is 3939 with 8 unique values.


## Confirming the catalogid provided by the function to check for 8 options

In [18]:
df_options_[df_options_['CatalogID'] == '3939']

CatalogID SKU OptionName                               Name  \
432      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
433      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
434      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
435      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
436      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
437      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
438      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
439      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
440      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
441      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
442      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
443      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
444      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
445      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
446      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
447      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
448      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
449      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
450      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
451      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
452      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
453      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
454      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
455      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
456      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
457      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
458      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
459      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   
460      3939  D8       Blue  Whelen Dominator 8 TIR3 Super LED   
461      3939  D8      White  Whelen Dominator 8 TIR3 Super LED   
462      3939  D8        Red  Whelen Dominator 8 TIR3 Super LED   
463      3939  D8      Amber  Whelen Dominator 8 TIR3 Super LED   

    OptionPartNumber OptionSetID  OptionSorting cat_w_option  
432                B        2987              2    3939_2987  
433                C        2987              2    3939_2987  
434                R        2987              2    3939_2987  
435                A        2987              2    3939_2987  
436                B        2988              3    3939_2988  
437                C        2988              3    3939_2988  
438                R        2988              3    3939_2988  
439                A        2988              3    3939_2988  
440                B        2989              4    3939_2989  
441                C        2989              4    3939_2989  
442                R        2989              4    3939_2989  
443                A        2989              4    3939_2989  
444                B        2990              5    3939_2990  
445                C        2990              5    3939_2990  
446                R        2990              5    3939_2990  
447                A        2990              5    3939_2990  
448                B        2991              6    3939_2991  
449                C        2991              6    3939_2991  
450                R        2991              6    3939_2991  
451                A        2991              6    3939_2991  
452                B        2992              7    3939_2992  
453                C        2992              7    3939_2992  
454                R        2992              7    3939_2992  
455                A        2992              7    3939_2992  
456                B        2993              8    3939_2993  
457                C        2993              8    3939_2993  
458                R        2993              8    3939_2993  
459                A        

## Checking for unique SKUs - ensuring no errors to avoid mismatch with website

In [19]:
df_options_[df_options_['OptionPartNumber'] =='']['SKU'].unique()

array(['M4B6', 'SA315U', 'SA350M', 'SC-550', 'SLWIC3', 'SLPMM',
       'IW0BAAAA', 'C-AS-840-11', 'C-EB-', 'HOWLER', 'F4N', 'ES100C',
       'ULB24', 'C-1800', 'C-1810', 'C-2410', 'R316', 'R416', 'ENFLB',
       'C-4014', 'MCRNS', 'LINSV2', 'GALAXYELITE', 'F4M', 'C-4015',
       'C-VS-0412-CAPR-1', 'SA350MH', 'CCSRN', 'LF35', 'FS', 'TCRWXSOLO',
       'TCRWXDUO', 'GMP-F150-2015', 'OELS', 'GALAXYELITE-PURPLE', 'PCP3',
       'C-', 'FS-8816', 'FS-0416', 'FS-0616', 'FS-0816', 'C6T', 'C6BTT',
       'C7T', 'C7BTT', 'C9T', 'C9BTT', 'ENRLB', 'TRIOFST-', 'EMPLB',
       'S-3017', 'C-AP-0325-1', 'C-AP-0625-1', 'C-AP-0645-1',
       'POLICE-ALGT', 'C6BU', 'C7BU', 'PLOWPKG4', 'PF200', 'PCH1', 'PCH2',
       'PHANTOM', 'E3', 'E6', 'E62', 'E64', 'E68', 'E66', 'PF200R',
       'PF200H', 'PF200S17', 'SHADOW', 'CY', 'SNM-PROMOTSHIRT',
       'Q-0420\t', 'Q', '50LENSL', 'C399', 'ARGES', 'CHOWLER', 'ION',
       'WION', 'XON', 'WXON', 'C399R', 'RLNT48', 'INTGPROMO', 'VALRPROMO',
       'ILSPROMO-SIF', 

# Test Run

Here we are checking the create_combo_dataframes function to check that it works on the sample dataframe, taken from the master list. 

In [20]:
df_test = df_options_.iloc[:106]
df_test.tail()

CatalogID SKU              OptionName                         Name  \
104      3495  90                     Red  Whelen 900 Series Super LED   
105      3495  90     Blue with Blue Lens  Whelen 900 Series Super LED   
106      3495  90       Red With Red Lens  Whelen 900 Series Super LED   
107      3495  90  Amber with Amber Lens   Whelen 900 Series Super LED   
108      3495  90  Green with Green Lens   Whelen 900 Series Super LED   

    OptionPartNumber OptionSetID  OptionSorting cat_w_option  
104           RR5FCR        2479              1    3495_2479  
105           BB5FBR        2479              1    3495_2479  
106           RR5FRR        2479              1    3495_2479  
107           AA5FAR        2479              1    3495_2479  
108           GG5FGR        2479              1    3495_2479

In [21]:
create_combo_dataframes(df_test)


CatalogID ProductID             SKU  \
0       3474   11.1002      11.1002.PM   
1       3474   11.1002     11.1002.GSM   
0       3475    11.120  11.1200.008-BB   
1       3475    11.120  11.1200.008-RB   
2       3475    11.120  11.1200.008-BC   
3       3475    11.120  11.1200.008-RC   
4       3475    11.120  11.1200.008-AC   
5       3475    11.120  11.1200.008-GG   
6       3475    11.120  11.1200.008-RR   
7       3475    11.120  11.1200.008-AA   
8       3475    11.120  11.1200.008-RA   
9       3475    11.120      11.1200-BB   
10      3475    11.120      11.1200-RB   
11      3475    11.120      11.1200-BC   
12      3475    11.120      11.1200-RC   
13      3475    11.120      11.1200-AC   
14      3475    11.120      11.1200-GG   
15      3475    11.120      11.1200-RR   
16      3475    11.120      11.1200-AA   
17      3475    11.120      11.1200-RA   
0       3479         2        2EB00ZCR   
1       3479         2        2EC00ZCR   
2       3479         2        2ER00ZCR   
3       3479         2        2EA00ZCR   
4       3479         2        2EJ00ZCR   
5       3479         2        2EA00ZAR   
6       3479         2        2EB00ZBR   
7       3479         2        2ER00ZRR   
8       3479         2        2FB00ZCR   
9       3479         2        2FC00ZCR   
10      3479         2        2FR00ZCR   
11      3479         2        2FA00ZCR   
12      3479         2        2FJ00ZCR   
13      3479         2        2FA00ZAR   
14      3479         2        2FB00ZBR   
15      3479         2        2FR00ZRR   
0       3483        68  68-1183725-1SB   
1       3483        68  68-1183725-2SB   
2       3483        68  68-1183725-3SB   
3       3483        68  68-1183725-5SB   
4       3483        68  68-1183726-1SB   
5       3483        68  68-1183726-2SB   
6       3483        68  68-1183726-3SB   
7       3483        68  68-1183726-5SB   
8       3483        68  68-3183725-1SB   
9       3483        68  68-3183725-2SB   
10      3483        68  68-3183725-3SB   
11      3483        68  68-3183725-5SB   
12      3483        68  68-3183726-1SB   
13      3483        68  68-3183726-2SB   
14      3483        68  68-3183726-3SB   
15      3483        68  68-3183726-5SB   
16      3483        68  68-1183725-1SB   
17      3483        68  68-1183725-2SB   
18      3483        68  68-1183725-3SB   
19      3483        68  68-1183725-5SB   
20      3483        68  68-1183726-1SB   
21      3483        68  68-1183726-2SB   
22      3483        68  68-1183726-3SB   
23      3483        68  68-1183726-5SB   
0       3485        50        50AA2ZCR   
1       3485        50        50CR2ZCR   
2       3485        50        50CC2ZCR   
3       3485        50        50RR2ZCR   
4       3485        50        50BB2ZCR   
5       3485        50        50RB2ZCR   
0       3486        50        50BB3ZCR   
1       3486        50        50CR3ZCR   
2       3486        50        50CC3ZCR   
3       3486        50        50RR3ZCR   
4       3486        50        50AA3ZCR   
5       3486        50        50RB3ZCR   
6       3486        50        50AC3ZCR   
7       3486        50        50RA3ZCR   
0       3487        68  68-1963237-50A   
1       3487        68  68-1963237-10A   
2       3487        68  68-1963237-20A   
3       3487        68  68-1963237-30A   
4       3487        68  68-1963583-50A   
5       3487        68  68-1963583-10A   
6       3487        68  68-1963583-20A   
7       3487        68  68-1963583-30A   
8       3487        68  68-1963583-50A   
9       3487        68  68-1963583-10A   
10      3487        68  68-1963583-20A   
11      3487        68  68-1963583-30A   
12      3487        68  68-1963583-50A   
13      3487        68  68-1963583-10A   
14      3487        68  68-1963583-20A   
15      3487        68  68-1963583-30A   
0       3489        60        60C02FCR   
1       3489        60        60A02FCR   
2       3489        60        60BR6FCR   
3       3489        60        60AC6FCR   
4       3489 

## Initial Test Run - Success

After checking through the test dataframe we see that the SKUs created were a match with the website. This is also after checks done in the API as well as the store website. 

## Run on entire SKU list

In [22]:
Master_SKU_list = create_combo_dataframes(df_options_)

## Checking number of rows after run

In [23]:
len(Master_SKU_list)

279997

Length of dataframe is 279,997 SKUs. Client has approved this length.

In [24]:
Master_SKU_list.head(20)

CatalogID ProductID             SKU                                  Name  \
0       3474   11.1002      11.1002.PM  Sho Me Universal On/Off/Flash Switch   
1       3474   11.1002     11.1002.GSM  Sho Me Universal On/Off/Flash Switch   
0       3475    11.120  11.1200.008-BB  Sho Me Low Profile LED Mini Lightbar   
1       3475    11.120  11.1200.008-RB  Sho Me Low Profile LED Mini Lightbar   
2       3475    11.120  11.1200.008-BC  Sho Me Low Profile LED Mini Lightbar   
3       3475    11.120  11.1200.008-RC  Sho Me Low Profile LED Mini Lightbar   
4       3475    11.120  11.1200.008-AC  Sho Me Low Profile LED Mini Lightbar   
5       3475    11.120  11.1200.008-GG  Sho Me Low Profile LED Mini Lightbar   
6       3475    11.120  11.1200.008-RR  Sho Me Low Profile LED Mini Lightbar   
7       3475    11.120  11.1200.008-AA  Sho Me Low Profile LED Mini Lightbar   
8       3475    11.120  11.1200.008-RA  Sho Me Low Profile LED Mini Lightbar   
9       3475    11.120      11.1200-BB  Sho Me Low Profile LED Mini Lightbar   
10      3475    11.120      11.1200-RB  Sho Me Low Profile LED Mini Lightbar   
11      3475    11.120      11.1200-BC  Sho Me Low Profile LED Mini Lightbar   
12      3475    11.120      11.1200-RC  Sho Me Low Profile LED Mini Lightbar   
13      3475    11.120      11.1200-AC  Sho Me Low Profile LED Mini Lightbar   
14      3475    11.120      11.1200-GG  Sho Me Low Profile LED Mini Lightbar   
15      3475    11.120      11.1200-RR  Sho Me Low Profile LED Mini Lightbar   
16      3475    11.120      11.1200-AA  Sho Me Low Profile LED Mini Lightbar   
17      3475    11.120      11.1200-RA  Sho Me Low Profile LED Mini Lightbar   

                       Labels  
0          Positive Momentary  
1   Negative/Ground Momentary  
0               Magnetic_Blue  
1           Magnetic_Red/Blue  
2         Magnetic_Blue/White  
3          Magnetic_Red/White  
4        Magnetic_Amber/White  
5              Magnetic_Green  
6                Magnetic_Red  
7              Magnetic_Amber  
8          Magnetic_Red/Amber  
9              Permanent_Blue  
10         Permanent_Red/Blue  
11       Permanent_Blue/White  
12        Permanent_Red/White  
13      Permanent_Amber/White  
14            Permanent_Green  
15              Permanent_Red  
16            Permanent_Amber  
17        Permanent_Red/Amber

## Manual Checks

In the dataframe we checked for the string '--' to see if there were any missing partnumbers that may have caused some SKUs to come out of order. These 5 products below were the ones that were found. The dataframe was adjusted and the product SKUs now match the website. 

In [25]:
Master_SKU_list[Master_SKU_list['SKU'].str.contains('--')].head(20)
#check F4N
#check LegacyDUO
#check LegacySOLO
#check LIBERTYIISOLO
#check SOLOFST-


CatalogID ProductID                       SKU  \
3        6463  SOLOFST-  SOLOFST-ISFW355--CONTACT   
9        6463  SOLOFST-  SOLOFST-ISFW35Z--CONTACT   
15       6463  SOLOFST-  SOLOFST-ISFW425--CONTACT   
21       6463  SOLOFST-  SOLOFST-ISFW426--CONTACT   
27       6463  SOLOFST-  SOLOFST-ISFW42X--CONTACT   
33       6463  SOLOFST-  SOLOFST-ISFW42Z--CONTACT   
39       6463  SOLOFST-  SOLOFST-ISFW434--CONTACT   
45       6463  SOLOFST-  SOLOFST-ISFW435--CONTACT   
51       6463  SOLOFST-  SOLOFST-ISFW438--CONTACT   
57       6463  SOLOFST-  SOLOFST-ISFW43Z--CONTACT   
63       6463  SOLOFST-  SOLOFST-ISFW444--CONTACT   
69       6463  SOLOFST-  SOLOFST-ISFW445--CONTACT   
75       6463  SOLOFST-  SOLOFST-ISFW448--CONTACT   
81       6463  SOLOFST-  SOLOFST-ISFW44Z--CONTACT   
87       6463  SOLOFST-  SOLOFST-ISFW455--CONTACT   
93       6463  SOLOFST-  SOLOFST-ISFW456--CONTACT   
99       6463  SOLOFST-  SOLOFST-ISFW45X--CONTACT   
105      6463  SOLOFST-  SOLOFST-ISFW45Z--CONTACT   
111      6463  SOLOFST-  SOLOFST-ISFW474--CONTACT   
117      6463  SOLOFST-  SOLOFST-ISFW475--CONTACT   

                                                 Name  \
3    Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
9    Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
15   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
21   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
27   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
33   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
39   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
45   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
51   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
57   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
63   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
69   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
75   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
81   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
87   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
93   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
99   Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
105  Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
111  Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   
117  Whelen Inner Edge FST WeCanX  SOLO™ Interior Bar   

                                                                                                  Labels  
3    WeCan Control *Requires WeCan Control Point or Device_Dodge Charger, 2018-2020, Five Lamps, Uppe...  
9    WeCan Control *Requires WeCan Control Point or Device_Dodge Charger, 2018-2020, Ten Lamps, Upper...  
15   WeCan Control *Requires WeCan Control Point or Device_Ram 1500, 2018 and Ram 1500 Classic, 2019,...  
21   WeCan Control *Requires WeCan Control Point or Device_Ram 1500, 2018 and Ram 1500 Classic, 2019,...  
27   WeCan Control *Requires WeCan Control Point or Device_Ram 1500, 2018 and Ram 1500 Classic, 2019,...  
33   WeCan Control *Requires WeCan Control Point or Device_Ram 1500, 2018 and Ram 1500 Classic, 2019,...  
39   WeCan Control *Requires WeCan Control Point or Device_Ford Fusion, 2017-2019, Four Lamps, Upper ...  
45   WeCan Control *Requires WeCan Control Point or Device_Ford Fusion, 2017-2019, Five Lamps, Upper ...  
51   WeCan Control *Requires WeCan Control Point or Device_Ford Fusion, 2017-2019, Eight Lamps, Upper...  
57   WeCan Control *Requires WeCan Control Point or Device_Ford Fusion, 2017-2019, Ten Lamps, Upper F...  
63   WeCan Control *Requires WeCan Control Point or Device_Dodge Durango, 2018-2020, Four Lamps, Uppe...  
69   WeCan Control *Requires WeCan Control Point or Device_Dodge Durango, 2018-2020, Five Lamps, Uppe...  
75   WeCan Control *Requires WeCan Control Point or Device_Dodge Durango, 2018-2020, Eight Lamps, Upp...  
81   WeCan Control *Requires WeCan Control Point or Device_Dodge Durango, 2018-2020, Ten Lamps, Upper...  
87   WeCan Control *Requires WeCan Control Point or Device_Chevy Tahoe/Suburban, 2017-2

# Checks on the following products:
- #check F4N **done**
- #check LegacyDUO **done**
- #check LegacySOLO **done**
- #check LIBERTYIISOLO **done**
- #check SOLOFST-	**done**

In [26]:
Master_SKU_list[Master_SKU_list['ProductID'] == 'F4N'].head(20)

CatalogID ProductID                SKU  \
0       4136       F4N   F4N2-MKEZ95-VLED   
1       4136       F4N   F4N2-MKEZ95-VLED   
2       4136       F4N   F4N2-MKEZ95-VLED   
3       4136       F4N   F4N2-MKEZ95-QLED   
4       4136       F4N   F4N2-MKEZ95-QLED   
5       4136       F4N   F4N2-MKEZ95-QLED   
6       4136       F4N   F4N2-MKEZ96-VLED   
7       4136       F4N   F4N2-MKEZ96-VLED   
8       4136       F4N   F4N2-MKEZ96-VLED   
9       4136       F4N   F4N2-MKEZ96-QLED   
10      4136       F4N   F4N2-MKEZ96-QLED   
11      4136       F4N   F4N2-MKEZ96-QLED   
12      4136       F4N  F4N2-MKEZ101-VLED   
13      4136       F4N  F4N2-MKEZ101-VLED   
14      4136       F4N  F4N2-MKEZ101-VLED   
15      4136       F4N  F4N2-MKEZ101-QLED   
16      4136       F4N  F4N2-MKEZ101-QLED   
17      4136       F4N  F4N2-MKEZ101-QLED   
18      4136       F4N   F4N2-MKEZ45-VLED   
19      4136       F4N   F4N2-MKEZ45-VLED   

                                              Name  \
0   Whelen NFPA Edge Freedom IV Super LED Lightbar   
1   Whelen NFPA Edge Freedom IV Super LED Lightbar   
2   Whelen NFPA Edge Freedom IV Super LED Lightbar   
3   Whelen NFPA Edge Freedom IV Super LED Lightbar   
4   Whelen NFPA Edge Freedom IV Super LED Lightbar   
5   Whelen NFPA Edge Freedom IV Super LED Lightbar   
6   Whelen NFPA Edge Freedom IV Super LED Lightbar   
7   Whelen NFPA Edge Freedom IV Super LED Lightbar   
8   Whelen NFPA Edge Freedom IV Super LED Lightbar   
9   Whelen NFPA Edge Freedom IV Super LED Lightbar   
10  Whelen NFPA Edge Freedom IV Super LED Lightbar   
11  Whelen NFPA Edge Freedom IV Super LED Lightbar   
12  Whelen NFPA Edge Freedom IV Super LED Lightbar   
13  Whelen NFPA Edge Freedom IV Super LED Lightbar   
14  Whelen NFPA Edge Freedom IV Super LED Lightbar   
15  Whelen NFPA Edge Freedom IV Super LED Lightbar   
16  Whelen NFPA Edge Freedom IV Super LED Lightbar   
17  Whelen NFPA Edge Freedom IV Super LED Lightbar   
18  Whelen NFPA Edge Freedom IV Super LED Lightbar   
19  Whelen NFPA Edge Freedom IV Super LED Lightbar   

                                                                                                 Labels  
0   55_Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”_Four Red C...  
1   55_Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”_Four Red C...  
2   55_Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”_Four Red C...  
3   55_Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”_Two Red Fr...  
4   55_Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”_Two Red Fr...  
5   55_Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”_Two Red Fr...  
6   55_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_Four Red Corners and Four Front (2 Red/2 Whi...  
7   55_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_Four Red Corners and Four Front (2 Red/2 Whi...  
8   55_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_Four Red Corners and Four Front (2 Red/2 Whi...  
9   55_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_Two Red Front Corners and Four Front (2 Red/...  
10  55_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_Two Red Front Corners and Four Front (2 Red/...  
11  55_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_Two Red Front Corners and Four Front (2 Red/...  
12  55_Silverado 1500 / 2021-2023 / 54"-56"_Four Red Corners and Four Front (2 Red/2 White)_55 Perma...  
13  55_Silverado 1500 / 2021-2023 / 54"-56"_Four Red Corners and Four Front (2 Red/2 White)_55 - 62 ...  
14  55_Silverado 1500 / 2021-2023 / 54"-56"_Four Red Corners and Four Front (2 Red/2 White)_62+ Two ...  
15  55_Silverado 1500 / 2021-2023 / 54"-56"_Two Red Front Corners and Four Front (2 Red/2 White) and...  
16  55_Silverado 1500 / 2021-2023 / 54"-56"_Two Red Front Corners and Four Front (2 Red/2 White) and...  
17  55_Silvera

# F4N - checked

In [27]:
df_options_[df_options_['CatalogID'] == '4311' ].head(20)

CatalogID        SKU  \
829      4311  LEGACYDUO   
830      4311  LEGACYDUO   
831      4311  LEGACYDUO   
832      4311  LEGACYDUO   
833      4311  LEGACYDUO   
834      4311  LEGACYDUO   
835      4311  LEGACYDUO   
836      4311  LEGACYDUO   
837      4311  LEGACYDUO   
838      4311  LEGACYDUO   
839      4311  LEGACYDUO   
840      4311  LEGACYDUO   
841      4311  LEGACYDUO   
842      4311  LEGACYDUO   
843      4311  LEGACYDUO   
844      4311  LEGACYDUO   
845      4311  LEGACYDUO   
846      4311  LEGACYDUO   
847      4311  LEGACYDUO   
848      4311  LEGACYDUO   

                                                                             OptionName  \
829                                                                           48 Inches   
830                                                                           54 Inches   
831                                                Red = Red/White Front Red/Amber Rear   
832                                            Blue = Blue/White Front. Blue/Amber Rear   
833     Red/Blue = Half Red Half Blue w/ White Front - Half Red Half Blue w/ Amber Rear   
834                                          Amber = Amber/White Front Amber/White Rear   
835                                          WeCan Control *Require WeCan Control Point   
836                                   WeCanX Control Point *Requires CORE Control Point   
837  Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”   
838                                    Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"   
839                                                Silverado 1500 / 2021-2023 / 54"-56"   
840                               Silverado Pickup / 1999-2006 (2007 Classic) / 54"-56"   
841                           Silverado 1500 / GMC 1500 Pickup / 2014-2018 / 54"-56"+B6   
842                                                         Tahoe / 2015-2020 / 54"-56"   
843              Tahoe / 2021-2023 / 54"-56" (Not for Use with Factory Roof Rack Rails)   
844                                                       Charger / 2005-2010 / 48"-50"   
845                                                       Charger / 2011-2023 / 48"-50"   
846                                                       Durango / 2011-2023 / 48"-50"   
847                                                       Durango / 2004-2010 / 48"-50"   
848                             Durango, Adjustable Leveling Foot / 2004-2010 / 48"-50"   

                           Name OptionPartNumber OptionSetID  OptionSorting  \
829  Whelen Legacy Duo Lightbar            8SP3-        4982              2   
830  Whelen Legacy Duo Lightbar            2SP3-        4982              2   
831  Whelen Legacy Duo Lightbar              -RT        5576              4   
832  Whelen Legacy Duo Lightbar              -BT        5576              4   
833  Whelen Legacy Duo Lightbar              -JT        5576              4   
834  Whelen Legacy Duo Lightbar            -FFFF        5576              4   
835  Whelen Legacy Duo Lightbar              -GB        5735              1   
836  Whelen Legacy Duo Lightbar              -EB        5735              1   
837  Whelen Legacy Duo Lightbar           MKEZ95        6034              3   
838  Whelen Legacy Duo Lightbar           MKEZ96        6034              3   
839  Whelen Legacy Duo Lightbar          MKEZ101        6034              3   
840  Whelen Legacy Duo Lightbar           MKEZ45        6034              3   
841  Whelen Legacy Duo Lightbar           MKEZ91        6034              3   
842  Whelen Legacy Duo Lightbar           MKEZ93        6034              3   
843  Whelen Legacy Duo Lightbar          MKAJ101        6034              3   
844  Whelen Legacy Duo Lightbar           MKEZ69        6034              3   
845  Whelen Legacy Duo Lightbar           MKEZ82        6034              3   
846  Whelen Legacy Duo Lightbar           MKEZ85        6034              3   
84

In [28]:
Master_SKU_list[Master_SKU_list['ProductID'] == 'LEGACYDUO'].head(20)

CatalogID  ProductID                            SKU  \
0       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ95-RT   
1       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ95-BT   
2       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ95-JT   
3       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ95-FFFF   
4       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ96-RT   
5       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ96-BT   
6       4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ96-JT   
7       4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ96-FFFF   
8       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ101-RT   
9       4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ101-BT   
10      4311  LEGACYDUO    LEGACYDUO-GB8SP3-MKEZ101-JT   
11      4311  LEGACYDUO  LEGACYDUO-GB8SP3-MKEZ101-FFFF   
12      4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ45-RT   
13      4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ45-BT   
14      4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ45-JT   
15      4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ45-FFFF   
16      4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ91-RT   
17      4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ91-BT   
18      4311  LEGACYDUO     LEGACYDUO-GB8SP3-MKEZ91-JT   
19      4311  LEGACYDUO   LEGACYDUO-GB8SP3-MKEZ91-FFFF   

                          Name  \
0   Whelen Legacy Duo Lightbar   
1   Whelen Legacy Duo Lightbar   
2   Whelen Legacy Duo Lightbar   
3   Whelen Legacy Duo Lightbar   
4   Whelen Legacy Duo Lightbar   
5   Whelen Legacy Duo Lightbar   
6   Whelen Legacy Duo Lightbar   
7   Whelen Legacy Duo Lightbar   
8   Whelen Legacy Duo Lightbar   
9   Whelen Legacy Duo Lightbar   
10  Whelen Legacy Duo Lightbar   
11  Whelen Legacy Duo Lightbar   
12  Whelen Legacy Duo Lightbar   
13  Whelen Legacy Duo Lightbar   
14  Whelen Legacy Duo Lightbar   
15  Whelen Legacy Duo Lightbar   
16  Whelen Legacy Duo Lightbar   
17  Whelen Legacy Duo Lightbar   
18  Whelen Legacy Duo Lightbar   
19  Whelen Legacy Duo Lightbar   

                                                                                                 Labels  
0   WeCan Control *Require WeCan Control Point_48 Inches_Colorado / 2015-2022 / Legacy®, Liberty™ II...  
1   WeCan Control *Require WeCan Control Point_48 Inches_Colorado / 2015-2022 / Legacy®, Liberty™ II...  
2   WeCan Control *Require WeCan Control Point_48 Inches_Colorado / 2015-2022 / Legacy®, Liberty™ II...  
3   WeCan Control *Require WeCan Control Point_48 Inches_Colorado / 2015-2022 / Legacy®, Liberty™ II...  
4   WeCan Control *Require WeCan Control Point_48 Inches_Suburban (Roof Rack Mount) / 2015-2020 / 55...  
5   WeCan Control *Require WeCan Control Point_48 Inches_Suburban (Roof Rack Mount) / 2015-2020 / 55...  
6   WeCan Control *Require WeCan Control Point_48 Inches_Suburban (Roof Rack Mount) / 2015-2020 / 55...  
7   WeCan Control *Require WeCan Control Point_48 Inches_Suburban (Roof Rack Mount) / 2015-2020 / 55...  
8   WeCan Control *Require WeCan Control Point_48 Inches_Silverado 1500 / 2021-2023 / 54"-56"_Red = ...  
9   WeCan Control *Require WeCan Control Point_48 Inches_Silverado 1500 / 2021-2023 / 54"-56"_Blue =...  
10  WeCan Control *Require WeCan Control Point_48 Inches_Silverado 1500 / 2021-2023 / 54"-56"_Red/Bl...  
11  WeCan Control *Require WeCan Control Point_48 Inches_Silverado 1500 / 2021-2023 / 54"-56"_Amber ...  
12  WeCan Control *Require WeCan Control Point_48 Inches_Silverado Pickup / 1999-2006 (2007 Classic)...  
13  WeCan Control *Require WeCan Control Point_48 Inches_Silverado Pickup / 1999-2006 (2007 Classic)...  
14  WeCan Control *Require WeCan Control Point_48 Inches_Silverado Pickup / 1999-2006 (2007 Classic)...  
15  WeCan Control *Require WeCan Control Point_48 Inches_Silverado Pickup / 1999-2006 (2007 Classic)...  
16  WeCan Control *Require WeCan Control Point_48 Inches_Silverado 1500 / GMC 1500 Pickup / 2014-201...  
17  WeCan Control *Require WeCan Control Point_48 Inches_Silverado 1500 / GMC 1500 Pickup / 2014-201...  
18  WeCan Control *Require WeCan Control Point_48 Inche

# LegacyDUO - checked


In [29]:
df_options_[df_options_['CatalogID'] == '4413'].head(20)

CatalogID         SKU  \
965      4413  LEGACYSOLO   
966      4413  LEGACYSOLO   
967      4413  LEGACYSOLO   
968      4413  LEGACYSOLO   
969      4413  LEGACYSOLO   
970      4413  LEGACYSOLO   
971      4413  LEGACYSOLO   
972      4413  LEGACYSOLO   
973      4413  LEGACYSOLO   
974      4413  LEGACYSOLO   
975      4413  LEGACYSOLO   
976      4413  LEGACYSOLO   
977      4413  LEGACYSOLO   
978      4413  LEGACYSOLO   
979      4413  LEGACYSOLO   
980      4413  LEGACYSOLO   
981      4413  LEGACYSOLO   
982      4413  LEGACYSOLO   
983      4413  LEGACYSOLO   
984      4413  LEGACYSOLO   

                                                                             OptionName  \
965                                                                Half Red / Half Blue   
966                                                                               Amber   
967                                                                                Blue   
968                                                                                 Red   
969                                                                           48 Inches   
970                                                                           54 Inches   
971                                          WeCan Control *Require WeCan Control Point   
972                                       WeCanX Conttrol *Requires CORE Control Point    
973  Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”   
974                                    Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"   
975                                                Silverado 1500 / 2021-2023 / 54"-56"   
976                               Silverado Pickup / 1999-2006 (2007 Classic) / 54"-56"   
977                           Silverado 1500 / GMC 1500 Pickup / 2014-2018 / 54"-56"+B6   
978                                                         Tahoe / 2015-2020 / 54"-56"   
979              Tahoe / 2021-2023 / 54"-56" (Not for Use with Factory Roof Rack Rails)   
980                                                       Charger / 2005-2010 / 48"-50"   
981                                                       Charger / 2011-2023 / 48"-50"   
982                                                       Durango / 2011-2023 / 48"-50"   
983                                                       Durango / 2004-2010 / 48"-50"   
984                             Durango, Adjustable Leveling Foot / 2004-2010 / 48"-50"   

                            Name OptionPartNumber OptionSetID  OptionSorting  \
965  Whelen Legacy SOLO Lightbar               -J        3346              4   
966  Whelen Legacy SOLO Lightbar               -A        3346              4   
967  Whelen Legacy SOLO Lightbar               -B        3346              4   
968  Whelen Legacy SOLO Lightbar               -R        3346              4   
969  Whelen Legacy SOLO Lightbar            8SP1-        4977              2   
970  Whelen Legacy SOLO Lightbar            2SP1-        4977              2   
971  Whelen Legacy SOLO Lightbar              -GS        5734              1   
972  Whelen Legacy SOLO Lightbar              -ES        5734              1   
973  Whelen Legacy SOLO Lightbar           MKEZ95        6032              3   
974  Whelen Legacy SOLO Lightbar           MKEZ96        6032              3   
975  Whelen Legacy SOLO Lightbar          MKEZ101        6032              3   
976  Whelen Legacy SOLO Lightbar           MKEZ45        6032              3   
977  Whelen Legacy SOLO Lightbar           MKEZ91        6032              3   
978  Whelen Legacy SOLO Lightbar           MKEZ93        6032              3   
979  Whelen Legacy SOLO Lightbar          MKAJ101        6032              3   
980  Whelen Legacy SOLO Lightbar           MKEZ69        6032              3   
981  Whelen Legacy SOLO Lightbar           MKEZ82        6032              3   
982  Whelen Legacy SOLO Lightbar          

In [30]:
Master_SKU_list[Master_SKU_list['ProductID'] == 'LEGACYSOLO'].head(100)

CatalogID   ProductID                          SKU  \
0       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ95-J   
1       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ95-A   
2       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ95-B   
3       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ95-R   
4       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ96-J   
5       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ96-A   
6       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ96-B   
7       4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ96-R   
8       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ101-J   
9       4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ101-A   
10      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ101-B   
11      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ101-R   
12      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ45-J   
13      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ45-A   
14      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ45-B   
15      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ45-R   
16      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ91-J   
17      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ91-A   
18      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ91-B   
19      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ91-R   
20      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ93-J   
21      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ93-A   
22      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ93-B   
23      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ93-R   
24      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKAJ101-J   
25      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKAJ101-A   
26      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKAJ101-B   
27      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKAJ101-R   
28      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ69-J   
29      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ69-A   
30      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ69-B   
31      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ69-R   
32      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ82-J   
33      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ82-A   
34      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ82-B   
35      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ82-R   
36      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ85-J   
37      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ85-A   
38      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ85-B   
39      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ85-R   
40      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ62-J   
41      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ62-A   
42      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ62-B   
43      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ62-R   
44      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKAJ62-J   
45      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKAJ62-A   
46      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKAJ62-B   
47      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKAJ62-R   
48      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ86-J   
49      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ86-A   
50      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ86-B   
51      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ86-R   
52      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ102-J   
53      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ102-A   
54      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ102-B   
55      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ102-R   
56      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ107-J   
57      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ107-A   
58      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ107-B   
59      4413  LEGACYSOLO  LEGACYSOLO-GS8SP1-MKEZ107-R   
60      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ89-J   
61      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ89-A   
62      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ89-B   
63      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ89-R   
64      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ38-J   
65      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ38-A   
66      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ38-B   
67      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ38-R   
68      4413  LEGACYSOLO   LEGACYSOLO-GS8SP1-MKEZ98-J   
69      4413 

# Legacy Solo - Checked

In [31]:
df_options_[df_options_['CatalogID'] == '6174'].head(20)

CatalogID            SKU  \
2438      6174  LIBERTYIISOLO   
2439      6174  LIBERTYIISOLO   
2440      6174  LIBERTYIISOLO   
2441      6174  LIBERTYIISOLO   
2442      6174  LIBERTYIISOLO   
2443      6174  LIBERTYIISOLO   
2444      6174  LIBERTYIISOLO   
2445      6174  LIBERTYIISOLO   
2446      6174  LIBERTYIISOLO   
2447      6174  LIBERTYIISOLO   
2448      6174  LIBERTYIISOLO   
2449      6174  LIBERTYIISOLO   
2450      6174  LIBERTYIISOLO   
2451      6174  LIBERTYIISOLO   
2452      6174  LIBERTYIISOLO   
2453      6174  LIBERTYIISOLO   
2454      6174  LIBERTYIISOLO   
2455      6174  LIBERTYIISOLO   
2456      6174  LIBERTYIISOLO   
2457      6174  LIBERTYIISOLO   

                                                                              OptionName  \
2438                                                                                  48   
2439                                                                                  54   
2440                                                                               Amber   
2441                                                                                Blue   
2442                                                                                 Red   
2443                                                                            Red/Blue   
2444                              Contact Me for Colors (This may come with an upcharge)   
2445                                          WeCan Control *Require WeCan Control Point   
2446                                   WeCanX Control Point *Requires CORE Control Point   
2447  Colorado / 2015-2022 / Legacy®, Liberty™ II and Freedom® IV 48”-50” or Justice 55”   
2448                                    Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"   
2449                                                Silverado 1500 / 2021-2023 / 54"-56"   
2450                               Silverado Pickup / 1999-2006 (2007 Classic) / 54"-56"   
2451                           Silverado 1500 / GMC 1500 Pickup / 2014-2018 / 54"-56"+B6   
2452                                                         Tahoe / 2015-2020 / 54"-56"   
2453              Tahoe / 2021-2023 / 54"-56" (Not for Use with Factory Roof Rack Rails)   
2454                                                       Charger / 2005-2010 / 48"-50"   
2455                                                       Charger / 2011-2023 / 48"-50"   
2456                                                       Durango / 2011-2023 / 48"-50"   
2457                                                       Durango / 2004-2010 / 48"-50"   

                                 Name OptionPartNumber OptionSetID  \
2438  Whelen Liberty II SOLO Lightbar               8-        5598   
2439  Whelen Liberty II SOLO Lightbar               2-        5598   
2440  Whelen Liberty II SOLO Lightbar            -AAAA        5599   
2441  Whelen Liberty II SOLO Lightbar            -BBBB        5599   
2442  Whelen Liberty II SOLO Lightbar            -RRRR        5599   
2443  Whelen Liberty II SOLO Lightbar            -BRBR        5599   
2444  Whelen Liberty II SOLO Lightbar            -Call        5599   
2445  Whelen Liberty II SOLO Lightbar              -IW        5736   
2446  Whelen Liberty II SOLO Lightbar              -BW        5736   
2447  Whelen Liberty II SOLO Lightbar           MKEZ95        6033   
2448  Whelen Liberty II SOLO Lightbar           MKEZ96        6033   
2449  Whelen Liberty II SOLO Lightbar          MKEZ101        6033   
2450  Whelen Liberty II SOLO Lightbar           MKEZ45        6033   
2451  Whelen Liberty II SOLO Lightbar           MKEZ91        6033   
2452  Whelen Liberty II SOLO Lightbar           MKEZ93        6033   
2453  Whelen Liberty II SOLO Lightbar          MKAJ101        6033   
2454  Whelen Liberty II SOLO Lightbar           MKEZ69        6033   
2455  Whelen Liberty II SOLO Lightbar           MKEZ82        6033   
2456  Whelen Liberty II SOLO Lightbar           

In [32]:
Master_SKU_list[Master_SKU_list['ProductID'] == 'LIBERTYIISOLO'].head(20)

CatalogID      ProductID                             SKU  \
0       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-AAAA   
1       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-BBBB   
2       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-RRRR   
3       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-BRBR   
4       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ95-Call   
5       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-AAAA   
6       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-BBBB   
7       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-RRRR   
8       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-BRBR   
9       6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ96-Call   
10      6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-AAAA   
11      6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-BBBB   
12      6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-RRRR   
13      6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-BRBR   
14      6174  LIBERTYIISOLO  LIBERTYIISOLO-IW8-MKEZ101-Call   
15      6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-AAAA   
16      6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-BBBB   
17      6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-RRRR   
18      6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-BRBR   
19      6174  LIBERTYIISOLO   LIBERTYIISOLO-IW8-MKEZ45-Call   

                               Name  \
0   Whelen Liberty II SOLO Lightbar   
1   Whelen Liberty II SOLO Lightbar   
2   Whelen Liberty II SOLO Lightbar   
3   Whelen Liberty II SOLO Lightbar   
4   Whelen Liberty II SOLO Lightbar   
5   Whelen Liberty II SOLO Lightbar   
6   Whelen Liberty II SOLO Lightbar   
7   Whelen Liberty II SOLO Lightbar   
8   Whelen Liberty II SOLO Lightbar   
9   Whelen Liberty II SOLO Lightbar   
10  Whelen Liberty II SOLO Lightbar   
11  Whelen Liberty II SOLO Lightbar   
12  Whelen Liberty II SOLO Lightbar   
13  Whelen Liberty II SOLO Lightbar   
14  Whelen Liberty II SOLO Lightbar   
15  Whelen Liberty II SOLO Lightbar   
16  Whelen Liberty II SOLO Lightbar   
17  Whelen Liberty II SOLO Lightbar   
18  Whelen Liberty II SOLO Lightbar   
19  Whelen Liberty II SOLO Lightbar   

                                                                                                 Labels  
0   WeCan Control *Require WeCan Control Point_48_Colorado / 2015-2022 / Legacy®, Liberty™ II and Fr...  
1   WeCan Control *Require WeCan Control Point_48_Colorado / 2015-2022 / Legacy®, Liberty™ II and Fr...  
2   WeCan Control *Require WeCan Control Point_48_Colorado / 2015-2022 / Legacy®, Liberty™ II and Fr...  
3   WeCan Control *Require WeCan Control Point_48_Colorado / 2015-2022 / Legacy®, Liberty™ II and Fr...  
4   WeCan Control *Require WeCan Control Point_48_Colorado / 2015-2022 / Legacy®, Liberty™ II and Fr...  
5   WeCan Control *Require WeCan Control Point_48_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_A...  
6   WeCan Control *Require WeCan Control Point_48_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_Blue  
7    WeCan Control *Require WeCan Control Point_48_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_Red  
8   WeCan Control *Require WeCan Control Point_48_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_R...  
9   WeCan Control *Require WeCan Control Point_48_Suburban (Roof Rack Mount) / 2015-2020 / 55"-60"_C...  
10             WeCan Control *Require WeCan Control Point_48_Silverado 1500 / 2021-2023 / 54"-56"_Amber  
11              WeCan Control *Require WeCan Control Point_48_Silverado 1500 / 2021-2023 / 54"-56"_Blue  
12               WeCan Control *Require WeCan Control Point_48_Silverado 1500 / 2021-2023 / 54"-56"_Red  
13          WeCan Control *Require WeCan Control Point_48_Silverado 1500 / 2021-2023 / 54"-56"_Red/Blue  
14  WeCan Control *Require WeCan Control Point_48_Silverado 1500 / 2021-2023 / 54"-56"_Contact Me fo...  
15  WeCan Control *Require WeCan Control Point_48_Silverado Pickup / 1999-2006 (2007 Classic) / 54"-...  
16  WeCan Control *Require WeCan Control Point_48_Silvera

# LIBERTYIISOLO - Checked

In [33]:
Master_SKU_list.tail(100)

CatalogID ProductID              SKU  \
228      6983        RX   RX2SP3AMBEZ108   
229      6983        RX   RX2SP3AMBEZ105   
230      6983        RX   RX2SP3AMBAJ105   
231      6983        RX    RX2SP3AMKLP95   
232      6983        RX    RX2SP3AMKLP91   
233      6983        RX    RX2SP3AMKLP93   
234      6983        RX    RX2SP3AMKLP82   
235      6983        RX    RX2SP3AMBLP82   
236      6983        RX   RX2SP3AMBEZ101   
237      6983        RX    RX2SP3AMKEZ94   
238      6983        RX    RX2SP3AMBEZ94   
239      6983        RX    RX2SP3AMBAJ94   
240      6983        RX    RX2SP3AMKEZ95   
241      6983        RX    RX2SP3AMKEZ96   
242      6983        RX   RX2SP3AMKEZ101   
243      6983        RX    RX2SP3AMKEZ45   
244      6983        RX    RX2SP3AMKEZ91   
245      6983        RX    RX2SP3AMKEZ93   
246      6983        RX   RX2SP3AMKAJ101   
247      6983        RX    RX2SP3AMKEZ69   
248      6983        RX    RX2SP3AMKEZ82   
249      6983        RX    RX2SP3AMKEZ85   
250      6983        RX    RX2SP3AMKEZ62   
251      6983        RX    RX2SP3AMKAJ62   
252      6983        RX    RX2SP3AMKEZ86   
253      6983        RX   RX2SP3AMKEZ102   
254      6983        RX   RX2SP3AMKEZ107   
255      6983        RX    RX2SP3AMKEZ89   
256      6983        RX    RX2SP3AMKEZ38   
257      6983        RX    RX2SP3AMKEZ98   
258      6983        RX    RX2SP3AMKAJ98   
259      6983        RX    RX2SP3AMKEZ94   
260      6983        RX    RX2SP3AMKAJ94   
261      6983        RX    RX2SP3AMKEA94   
262      6983        RX   RX2SP3AMKEZ38L   
263      6983        RX    RX2SP3AMKEZ88   
264      6983        RX   RX2SP3AMKEZ108   
265      6983        RX   RX2SP3AMKEZ105   
266      6983        RX   RX2SP3AMKAJ105   
267      6983        RX    RX2SP3AMKEZ83   
268      6983        RX   RX2SP3AMKEZ104   
269      6983        RX   RX2SP3AMKAJ104   
270      6983        RX    RX2SP3AMKEZ99   
271      6983        RX   RX2SP3AMBAJ101   
272      6983        RX    RX2SP3AMBEZ82   
273      6983        RX    RX2SP3AMBEZ85   
274      6983        RX    RX2SP3AMKAJ85   
275      6983        RX    RX2SP3AMBAJ85   
276      6983        RX   RX2SP3AMBEZ108   
277      6983        RX   RX2SP3AMBEZ105   
278      6983        RX   RX2SP3AMBAJ105   
279      6983        RX    RX2SP3AMKLP95   
280      6983        RX    RX2SP3AMKLP91   
281      6983        RX    RX2SP3AMKLP93   
282      6983        RX    RX2SP3AMKLP82   
283      6983        RX    RX2SP3AMBLP82   
284      6983        RX   RX2SP3AMBEZ101   
285      6983        RX    RX2SP3AMKEZ94   
286      6983        RX    RX2SP3AMBEZ94   
287      6983        RX    RX2SP3AMBAJ94   
0        6992    S-3700          S-3700R   
1        6992    S-3700          S-3700W   
2        6992    S-3700          S-3700B   
3        6992    S-3700          S-3700A   
0        6993    S-4600          S-4600R   
1        6993    S-4600          S-4600W   
2        6993    S-4600          S-4600B   
3        6993    S-4600          S-4600A   
0        6994    S-7900          S-7900R   
1        6994    S-7900          S-7900W   
2        6994    S-7900          S-7900B   
3        6994    S-7900          S-7900A   
0        6995  BTT-7900       BTT-7900RR   
1        6995  BTT-7900         BTT-7900   
0        6996  BTT-3700       BTT-3700RR   
1        6996  BTT-3700         BTT-3700   
0        6997  BTT-4600       BTT-4600RR   
1        6997  BTT-4600         BTT-4600   
0        6998   TT-7900        TT-7900AA   
1        6998   TT-7900          TT-7900   
0        6999   TT-3700        TT-3700AA   
1        6999   TT-3700          TT-3700   
0        7000   TT-4600        TT-4600AA   
1        7000   TT-4600          TT-4600   
0        7024  SAK73ADP  SAK73ADPEB00ZCR   
1        7024  SAK73ADP  SAK73ADPEC00ZCR   
2        7024  SAK73ADP  SAK73ADPER00ZCR   
3        7024  SAK73ADP  SAK73ADPEA00ZCR   
4        7024  SAK73ADP  SAK73ADPEJ00ZCR   
5        7024  SAK73ADP  SAK73ADPEA00ZAR   


In [34]:
Master_SKU_list[Master_SKU_list['SKU'].str.endswith('.0')]['CatalogID'].unique()

array(['3500', '3794', '3980', '4087', '4089', '4278', '4471', '5352',
       '5411', '5441', '5543', '5613', '5614', '5745', '5890', '5899',
       '5947', '5957', '6031', '6032', '6037', '6038', '6041', '6042',
       '6043', '6292', '6297', '6337', '6366', '6388', '6406', '6408',
       '6667', '6826', '6950', '6979'], dtype=object)

In [35]:
Master_SKU_list[Master_SKU_list['SKU'].str.endswith('.0')].head(20)

CatalogID ProductID             SKU  \
0      3500     ALPHA        ALPHA1.0   
1      3500     ALPHA        ALPHA2.0   
2      3500     ALPHA        ALPHA3.0   
3      3500     ALPHA        ALPHA4.0   
4      3500     ALPHA        ALPHA5.0   
0      3794     PA300  PA300-690000.0   
1      3794     PA300  PA300-690010.0   
0      3980       WCC          WCC9.0   
1      3980       WCC         WCC92.0   
0      4087       905         90540.0   
1      4087       905          9053.0   
0      4089    NSP-11      NSP-1170.0   
1      4089    NSP-11      NSP-1172.0   
2      4089    NSP-11      NSP-1174.0   
0      4278     CCSRN        CCSRN3.0   
1      4278     CCSRN       CCSRN35.0   
2      4278     CCSRN       CCSRN36.0   
0      4471        PC           PC2.0   
1      4471        PC           PC3.0   
2      4471        PC           PC1.0   

                                                             Name  \
0                         Whelen Remote Switches for Alpha Sirens   
1                         Whelen Remote Switches for Alpha Sirens   
2                         Whelen Remote Switches for Alpha Sirens   
3                         Whelen Remote Switches for Alpha Sirens   
4                         Whelen Remote Switches for Alpha Sirens   
0                           Federal Signal PA300 Electronic Siren   
1                           Federal Signal PA300 Electronic Siren   
0                              Whelen WeCan Electronic Controller   
1                              Whelen WeCan Electronic Controller   
0                 Streamlight Survivor Right Angle Personal Light   
1                 Streamlight Survivor Right Angle Personal Light   
0                                  Bayco 2 In 1 Safety Flashlight   
1                                  Bayco 2 In 1 Safety Flashlight   
2                                  Bayco 2 In 1 Safety Flashlight   
0  Whelen CenCom Saphire Switchbox Siren Module with Control Head   
1  Whelen CenCom Saphire Switchbox Siren Module with Control Head   
2  Whelen CenCom Saphire Switchbox Siren Module with Control Head   
0                          Whelen Heavy Duty Power Control Switch   
1                          Whelen Heavy Duty Power Control Switch   
2                          Whelen Heavy Duty Power Control Switch   

                                                       Labels  
0                                              Alpha 1 Switch  
1                                              Alpha 2 Switch  
2                                              Alpha 3 Switch  
3                                              Alpha 4 Switch  
4                                              Alpha 5 Switch  
0                                                    100 Watt  
1                                                    200 Watt  
0               WCC9 (Justice and Freedom IV Wecan Bars Only)  
1  WCC92 (Liberty II and Legacy Solo and Duo Wecan Lightbars)  
0                             Alkaline Model (4 AA Batteries)  
1                          Rechargable Model (1 NiCd battery)  
0                        Red/Blue Flood with White Flashlight  
1                       Red/White Flood with White Flashlight  
2       Full White Flashing Amber Flood with White Flashlight  
0                                                      CCSRN3  
1                                                     CCSRN35  
2                                                     CCSRN36  
0                                                 PC-2 Switch  
1                                                 PC-3 Switch  
2                                                 PC-1 Switch

## Removing the '.0'

After checks, it was noticed that some of the SKUs came in with an extra '.0' at the end. This happened with SKUs ending in a numerical value. These were removed to complete the match. 

In [36]:
def remove_decimal(value):
    if value.endswith('.0'):
        return value[:-2]  # Remove the last two characters (.0)
    return value

# Apply the function to the column where you want to remove '.0'
Master_SKU_list['SKU'] = Master_SKU_list['SKU'].apply(remove_decimal)

In [37]:
Master_SKU_list[Master_SKU_list['SKU'].str.endswith('.0')]['CatalogID'].unique()

array([], dtype=object)

All '.0' parts of the SKUs were successfully removed. 

# Here, Master_SKU_List is exported to a CSV

In [38]:
Master_SKU_list.to_csv('Master_SKU_list.csv')

In [39]:
len(Master_SKU_list)

279997

In [40]:
Master_SKU_list.columns

Index(['CatalogID', 'ProductID', 'SKU', 'Name', 'Labels'], dtype='object')

In [41]:
Master_SKU_list.isnull().sum()

CatalogID    0
ProductID    0
SKU          0
Name         0
Labels       0
dtype: int64

# Conclusion

In this project, we have created a Master SKU List of all products from the Strobe's N More Website. 

This list was compiled and pushed to a CSV for the client, along with a written report and a list of products off the website. 

## Future Work


Client was informed via the written report that the API contains some issues of ordering that made manual checks necessary. Client was advised to make the changes in the API to overwrite for future Inventory issues. 